File: Triatlon-Java/TriatlonNetworking/src/main/java/rpcprotocol/TriatlonServiceRpcProxy.java
package rpcprotocol;

import domain.*;
import service.TriatlonObserverInterface;
import service.TriatlonServiceInterface;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class TriatlonServiceRpcProxy implements TriatlonServiceInterface {

    private String host;

    private int port;
    private TriatlonObserverInterface client;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private Socket connection;
    private BlockingQueue<Response> qresponses;
    private volatile boolean finished;

    public TriatlonServiceRpcProxy(String host, int port) {
        this.host = host;
        this.port = port;
        qresponses = new LinkedBlockingQueue<Response>();
    }

    /*@Override
    public synchronized Arbitru getAccount(String username, String parola, TriatlonObserverInterface client) throws Exception {
        initializeConnection();
        Request req = new Request.Builder().type(RequestType.GET_ARBITRU).data(new ArrayList<>(Arrays.asList(username, parola))).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            throw new Exception(err);
        }
        else {
            this.client = client;
            return (Arbitru) response.data();
        }
    }*/

    @Override
    public synchronized Arbitru login(ArbitruDTO arbitru, TriatlonObserverInterface client) throws Exception {
        initializeConnection();
        Request req = new Request.Builder().type(RequestType.LOGIN).data(arbitru).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            throw new Exception(err);
        }
        else {
            if (response.type() == ResponseType.OK) {
                this.client = client;
            }
        }
        return (Arbitru) response.data();
    }

    @Override
    public void logout(Arbitru arbitru, TriatlonObserverInterface client) throws Exception {
        Request req = new Request.Builder().type(RequestType.LOGOUT).data(arbitru).build();
        sendRequest(req);
        Response response = readResponse();
        closeConnection();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            System.out.println(err);
        }
    }

    @Override
    public List<ParticipantDTO> getParticipantiCuPunctaj(TriatlonObserverInterface client) throws Exception {
        Request req = new Request.Builder().type(RequestType.GET_PARTICIPANTI_CU_PUNCTAJ).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            System.out.println(err);
        }
        return (List<ParticipantDTO>) response.data();
    }

    @Override
    public Iterable<Proba> findAllProbe() throws Exception {
        Request req = new Request.Builder().type(RequestType.GET_PROBE).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            System.out.println(err);
        }
        return (Iterable<Proba>) response.data();
    }

    @Override
    public void addInscriere(Inscriere inscriere) throws Exception {
        Request req = new Request.Builder().type(RequestType.ADD_INSCRIERE).data(inscriere).build();
        sendRequest(req);

        Response response = readResponse();

        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            System.out.println(err);
        }
    }

    @Override
    public List<ParticipantDTO> getParticipantiDeLaProba(Proba proba, TriatlonObserverInterface client) throws Exception {
        Request req = new Request.Builder().type(RequestType.GET_PARTICIPANTI_DE_LA_PROBA).data(proba).build();
        sendRequest(req);
        Response response = readResponse();
        if (response.type() == ResponseType.ERROR) {
            String err = response.data().toString();
            closeConnection();
            System.out.println(err);
        }
        return (List<ParticipantDTO>) response.data();
    }

    private void sendRequest(Request request) throws Exception {
        try {
            //System.out.println(request);
            output.writeObject(request);
            output.flush();
        } catch (IOException e) {
            throw new Exception("Error sending object "+e);
        }

    }

    private Response readResponse(){
        Response response=null;
        try{
            response = qresponses.take();
            //System.out.println(response);
        } catch (InterruptedException e) {
            e.printStackTrace();

        }
        return response;
    }

    private void startReader(){
        Thread tw=new Thread(new ReaderThread());
        tw.start();
    }

    private void initializeConnection() {
        try {
            connection=new Socket(host,port);
            output=new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input=new ObjectInputStream(connection.getInputStream());
            finished=false;
            startReader();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private void closeConnection() {
        finished=true;
        try {
            input.close();
            output.close();
            connection.close();
            client=null;
        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private void handleUpdate(Response response){
        if(response.type() == ResponseType.UPDATE_PUNCTAJ){
            try{
                client.updatePunctaj();
            }catch (Exception e){
                e.printStackTrace();
            }
        }
    }

    private boolean isUpdate(Response response){
        return response.type() == ResponseType.UPDATE_PUNCTAJ;
    }

    private class ReaderThread implements Runnable{
        public void run() {
            while(!finished){
                try {
                    Object response=input.readObject();
                    System.out.println("response received " + response);
                    if(isUpdate((Response) response)){
                        handleUpdate((Response) response);
                    }
                    else {
                        try {
                            qresponses.put((Response) response);
                            System.out.println("Recieved: " + qresponses);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                } catch (IOException | ClassNotFoundException e) {
                    System.out.println("Reading error "+e);
                }
            }
        }
    }
}

File: Triatlon-C#/TriatlonNetworking/rpProtocols/TriatlonServiceRpcProxy.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using TriatlonModel;
using TriatlonServices;

namespace TriatlonNetworking.rpProtocols
{
    public class TriatlonServiceRpcProxy : TriatlonServiceInterface
    {
        private string host;
        private int port;

        private TriatlonObserverInterface client;
        private NetworkStream stream;
        private IFormatter formatter;
        private TcpClient connection;

        private Queue<Response> responses;
        private volatile bool finished;
        private EventWaitHandle _waitHandle;

        public TriatlonServiceRpcProxy(string host, int port)
        {
            this.host = host;
            this.port = port;
            responses = new Queue<Response>();
        }

        private void initializeConnection()
        {
            try
            {
                connection = new TcpClient(host, port);
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                finished = false;
                _waitHandle = new AutoResetEvent(false);
                startReader();
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }
        private void startReader()
        {
            Thread tw = new Thread(run);
            tw.Start();
        }

        private bool isUpdate(Response response)
        {
            return response.type == ResponseType.UPDATE_PUNCTAJ;
        }

        public virtual void run()
        {
            while (!finished)
            {
                try
                {
                    object response = (Response)formatter.Deserialize(stream);
                    if (isUpdate((Response)response))
                    {
                        handleUpdate((Response)response);
                    }
                    else
                    {
                        lock (responses)
                        {
                            responses.Enqueue((Response)response);
                        }
                        _waitHandle.Set();
                    }
                }
                catch (Exception ex)
                {
                    Console.WriteLine("Reading error " + ex);
                }
            }
        }

        private void handleUpdate(Response response)
        {
            if (response.type == ResponseType.UPDATE_PUNCTAJ)
            {
                try
                {
                    client.updatePunctaj();
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
        }


        private Response readResponse()
        {
            Response response = null;
            try
            {
                _waitHandle.WaitOne();
                lock (responses)
                {
                    response = responses.Dequeue();
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
            return response;
        }


        private void sendRequest(Request request)
        {
            try
            {
                formatter.Serialize(stream, request);
                stream.Flush();
            }
            catch (Exception e)
            {
                throw new Exception("Error sending object " + e);
            }
        }

        private void closeConnection()
        {
            finished = true;
            try
            {
                stream.Close();
                connection.Close();
                _waitHandle.Close();
                client = null;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }


        public virtual Arbitru login(ArbitruDTO arbitru, TriatlonObserverInterface client)
        {
            initializeConnection();
            Request request = new Request.Builder().type(RequestType.LOGIN).data(arbitru).build();
            sendRequest(request);
            Response response = readResponse();
            if(response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                closeConnection();
                throw new Exception(err);
            }
            this.client = client;
            return (Arbitru)response.data;
        }

        public void logout(Arbitru arbitru, TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.LOGOUT).data(arbitru).build();
            sendRequest(request);
            Response response = readResponse();
            closeConnection();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }

            MessageBox.Show("Logout realizat cu succes!");
        }

        public virtual List<ParticipantDTO> GetParticipantiCuPunctaj(TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.GET_PARTICIPANTI_CU_PUNCTAJ).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
            List<ParticipantDTO> participanti = (List<ParticipantDTO>)response.data;
            return participanti;
        }

        public virtual IEnumerable<Proba> FindAllProbe(TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.PROBE).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
            IEnumerable<Proba> probe = (IEnumerable<Proba>)response.data;
            return probe;
        }

        public void AddInscriere(Inscriere inscriere, TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.ADD_INSCRIERE).data(inscriere).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
        }

        public virtual List<ParticipantDTO> GetParticipantiDeLaProba(Proba proba, TriatlonObserverInterface client)
        {
            /*Request request = new Request.Builder().type(RequestType.GET_PARTICIPANTI_DE_LA_PROBA).data(proba).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
            Dictionary<Participant, long> participanti = (Dictionary<Participant, long>)response.data;
            return participanti;*/
            return null;
        }

        public IEnumerable<ParticipantDTO> GetParticipantiDTO(TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.Get_Participanti_DTO).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
            IEnumerable<ParticipantDTO> participanti = (IEnumerable<ParticipantDTO>)response.data;
            return participanti;
        }

        public IEnumerable<ParticipantDTO> GetParticipantiDTOprobaActuala(Proba proba, TriatlonObserverInterface client)
        {
            Request request = new Request.Builder().type(RequestType.Get_Participanti_DTO_probaActuala).data(proba).build();
            sendRequest(request);
            Response response = readResponse();
            if (response.type == ResponseType.ERROR)
            {
                string err = response.data.ToString();
                throw new Exception(err);
            }
            IEnumerable<ParticipantDTO> participanti = (IEnumerable<ParticipantDTO>)response.data;
            return participanti;
        }

        public Arbitru GetAccount(string username, string parola)
        {
            return null;
        }
    }
}
